File: 0001~0500/0018. 4Sum/main.cpp
/** 
 * Source: https://leetcode.com/problems/4sum/
 * Date: 2021/7/17
 * Skill: Two Pointers
 * Runtime: 116 ms, faster than 24.91% of C++ online submissions
 * Memory Usage: 12.9 MB, less than 83.39% of C++ online submissions
 * Time complexity: 
 * Space complexity: 
 * Constraints: 
 *      1 <= nums.length <= 200
 *      -10^9 <= nums[i], target <= 10^9
 *      * the solution set must not contain duplicate triplets.
 */

#include<iostream>
#include<vector>
#include<string>
using namespace std;


class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        vector<vector<int>> ans;
        for (int i = 0; i < n-3; i++) {
            if (i && nums[i] == nums[i-1])
                continue;
            for (int j = i+1; j < n-2; j++) {
                if (j != i+1 && nums[j] == nums[j-1])
                    continue;
                int l = j+1, r = n-1;
                while (l < r) {
                    if (nums[i] + nums[j] + nums[l] + nums[r] == target) {
                        ans.push_back({nums[i], nums[j], nums[l++], nums[r--]});
                        while (l < r && nums[l] == nums[l-1])
                            l++;
                        while (l < r && nums[r] == nums[r+1])
                            r--;
                    } else if (nums[i] + nums[j] + nums[l] + nums[r] > target) 
                        r--;
                    else
                        l++;
                }
            }
        }
        return ans;
    }
};

static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 0001~0500/0018. 4Sum/main.py
"""
Source: https://rb.gy/l5ky9
Date: 2023/8/4
Skill:
Ref:
Runtime: 801 ms, faster than 67.85%
Memory Usage: 16.3 MB, less than 91.70%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        res, sz = [], len(nums)
        for i in range(sz - 3):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            for j in range(i + 1, sz - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                left, right, cur_sum = j + 1, sz - 1, nums[i] + nums[j]
                while left < right:
                    if nums[left] + nums[right] + cur_sum == target:
                        res.append([nums[i], nums[j], nums[left], nums[right]])
                        left += 1
                        right -= 1
                        while left < right and nums[left] == nums[left - 1]:
                            left += 1
                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1
                    elif nums[left] + nums[right] + cur_sum < target:
                        left += 1
                    else:
                        right -= 1

        return res


if __name__ == "__main__":
    s = Solution()
